Sources/SwiftParser/Statements.swift

```diff
@@ -1038,4 +1066,39 @@ extension Parser.Lookahead {
     } while lookahead.at(.poundIf, .poundElseif, .poundElse) && lookahead.hasProgressed(&loopProgress)
     return lookahead.atStartOfSwitchCase()
   }
+  
+  mutating func atStartOfStatementBody() -> Bool {
+    
+    guard consume(if: .leftBrace) != nil else {
+      return false
+    }
+        
+    var loopProgress = LoopProgressCondition()
+    
+    var braces = 1
+
+    while !at(.endOfFile) && hasProgressed(&loopProgress) {
+      if consume(if: .rightBrace) != nil {
+        braces -= 1
+      } else {
+        consumeAnyToken()
+      }
+      if braces == 0 {
+        if self.at(.comma) {
+          return false
+        }
+        if !self.atStartOfLine, self.at(.leftParen) || self.at(.leftBrace) {
+          return false
+        }
+        return true
```

Comment:
Sorry that I’m in the “Let me go and break your code mood. The following is currently parsed as an expression but would now be parsed as a body

```swift
if true, { (x: () -> Void) in return true} != nil {
}
```

But I think this might be sufficiently stupid that we can ignore it.

CC @bnbarham @hamishknight @rintaro Would like to hear your opinions here as well.

Reply:
I'm experimenting with changing `atStartOfStatementBody()` to check for `true` cases instead. I think correct bodies are less error prone so maybe there's less risk of changing existing code.

```swift
mutating func atStartOfStatementBody() -> Bool {
  guard at(.leftBrace) else {
    return false
  }
  skipSingle()
  if self.at(.endOfFile) || self.at(.semicolon, .keyword(.else)) || self.at(.rightBrace, .rightParen) {
    return true
  }
  if self.atStartOfLine {
    return !withLookahead { $0.atStartOfStatementBody() } && !self.at(.binaryOperator)
  }
  return false
}
```

The following statements are all parsing the intended body:

```swift
if true, { print("body") }

if true, { print("body") };

if true, { print("body") } else { }

{ if true, { print("body") } }

( if true, { print("body") } )

if true, { true }, { print("body") }

if true, { true } { print("body") }

if true, { true }
{ print("body") }

if true, { (x: () -> Void) in return true} != nil { print("body") }

if true, { (x: () -> Void) in return true }
!= nil { print("body") }

```

What do you think? @ahoppen

Reply:
Yes, I think this is a lot safer and easier to follow.

I just went through the code, added some explanations (please verify that they are correct) and some questions as well

```swift
mutating func atStartOfStatementBody() -> Bool {
  guard at(.leftBrace) else {
    // Statement bodies always start with a '{'. If there is no '{', we can't be at the statement body
    return false
  }
  skipSingle()
  if self.at(.endOfFile) {
    // There's nothing else in the source file, which could be the statement body. So this must be it.
  }
  if self.at(.semicolon) {
    // We can't have a semicolon between the condition and the statement body. So this must be the statement body.
    return true
  }
  if self.at(.keyword(.else)) {
    // If the next token is an `else` keyword, this must be the statement body of an `if` statement since the condition cannot be followed by `else`.
    // Note that `guard` condition does not u se `atStartOfStatementBody`
    return true
  }
    
  if self.at(.rightBrace) {
    // What is this needed for? Do you have an explanation?
    return true
  }
  if self.at(.rightParen) {
    // What is this condition for? Do you have an explanation?
    return true
  }
  if self.atStartOfLine {
    // If this is followed by a newline, it is most likely a statement body. The only exceptions are
    if withLookahead({ $0.atStartOfStatementBody() }) {
      // If a statement body starts after this one. In this case we have something like
      // if takesTrailing { print("trailing closure") }
      // {
      //   print("if body")
      // }
      return false
    }
    if self.at(.binaryOperator) {
      // If we have a binary operator. An if/switch expression cannot be the left-hand side of an operator, so this cannot be the if body, eg.
      // if takesTrailing { print("trailing closure") }
      // != nil 
      // {
      //   print("body")
      // }
      return false
    }
    // Does this correctly handle something like the following?
    // if takesTrailing { print("trailing closure") }
    //   , true {
    //  
    // }

    // I think it might be safer if we could also somehow flip this into a whitelist instead of blacklist rule.
    // Ie. only return `true` if the next token is at the start of a line *and* satisfies some condition.
    return true
  }
  return false
}
```

Reply:
The comments are great! It’s okay to add them to the PR?

I use `at(.rightBrace)` for a case like `if true,  {  if true,  { }  }` and `at(. rightParen)` for if statements surrounded by parenthesis (It’s not valid code but without handling this there may be regression in diagnostics). 

I’m pretty sure we are safe for trailing closure because the function name will prevent the lookahead from going beyond `guard at(.leftBrace)`.

By “flip this into a whitelist” do you mean came up with a condition that returns true to assert that it’s the body? There are so many possibilities for the token at the start of a line that I can only think “is at the start of a line *and* does **not** satisfies some condition.”

Reply:
I'm a little confused 

> Which of my comments does this refer to?

I was refering to
```swift
if takesTrailing { print("trailing closure") }
, true {
 
}
```

I'm confused with your examples using trailing closure because in my tests they don't go past beyond  `guard at(.leftBrace)` because the function name, only closures like `if true, { print("closure") } { print("body") }` will.

The `.doStuff` is something returning a `Bool` right? Your example doesn't even compile today which took me by surprise 😮 It compiles though if the `.doStuff` is on the same line.

```swift
if takesTrailing { print("trailing") }
    .doStuff {
}
// ❌ Cannot convert value of type '(() -> Void) -> Foo' to expected condition type 'Bool'
// ❌ Reference to member 'doStuff' cannot be resolved without a contextual type

if takesTrailing { print("trailing") }.doStuff { // ✅
}
```

Reply:
> I'm confused with your examples using trailing closure because in my tests they don't go past beyond `guard at(.leftBrace)` because the function name, only closures like `if true, { print("closure") } { print("body") }` will.

Oh, I think my brain just switched modes and I started thinking about trailing closure disambiguation instead of first-level closures. Sorry about that. Disregard most things I said. The only meaningful examples that I have would be 

```swift
if true,
nil != { return true }
, true {
}
```

And silly postfix operators, which luckily need to be on the same line. But maybe good to add a test case for this as well.

```swift
postfix operator +++

postfix func +++(_ x: () -> Bool) -> Bool { true }

if true,
{ return true }+++ {
}
```